src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIMarsHandle.cs

```diff
@@ -55,6 +55,11 @@ internal sealed class SNIMarsHandle : SNIHandle
         /// </summary>
         public override void Dispose()
         {
+            // SendControlPacket will lock so make sure that it cannot deadlock by failing to enter the DemuxerLock
+            if (_connection != null && Monitor.IsEntered(_connection.DemuxerSync))
+            {
+                throw new InvalidOperationException("SNIMarsHandle.HandleRecieveComplete should be be called while holding the SNIMarsConnection.DemuxerSync because it can cause deadlocks");
```

Comment:
Do we need to clear the connection here before throwing an exception?

Reply:
Unknown, this is a new exception and may not even need to be an exception it could just be an assertion. I needed it in place during development to make sure i didn't accidentally call this while holding the lock. What do you think?

Reply:
I vote for assertion.

Reply:
Ok. done. Just bear in mind that the tests for this repo are never run in debug mode so assertions won't fire in testing.